### PR TITLE
ENH: fix numerical instability around zero in boxcox_llf

### PR DESCRIPTION
**Pull request by xuefeng-xu**
#### Reference issue

Follow up #19604

#### What does this implement/fix?

The computation in `boxcox_llf` is not numerically stable around lambda=0.
This PR modify the computation of the log variance term as follows.

$$\begin{aligned}
\ln\sigma^2_{\psi_{\text{BC}}(\lambda,x)}
&= \ln\text{Var}[(x^\lambda-1)/\lambda] && x \text{ is the sample data} \\
&= \ln\text{Var}(x^\lambda/\lambda) && \text{ Ignore the subtraction of a constant} \\
&= \ln\lambda^{-2}\text{Var}(x^\lambda) && \text{Factor out the division by } \lambda \\
&= -2\ln|\lambda|+\ln\text{Var}(x^\lambda) \\
\end{aligned}$$

#### Additional information

Comparasion of the computation in linear space, log space old (main branch), log space new (this PR) around lambda=0.

<div align=center>
    
</div>

<div align=center>
    
</div>

<div align=center>
    
</div>

<details>

<summary>code</summary>

```python
import numpy as np
from scipy.stats._morestats import _log_var
import matplotlib.pyplot as plt


def boxcox_llf_linear_space(lmb, data):
    data = np.asarray(data)
    N = data.shape[0]
    if N == 0:
        return np.nan
    logdata = np.log(data)

    if lmb == 0:
        logvar = np.log(np.var(logdata, axis=0))
    else:
        logvar = np.log(np.var(data**lmb / lmb, axis=0))

    return (lmb - 1) * np.sum(logdata, axis=0) - N/2 * logvar


def boxcox_llf_log_space_old(lmb, data):
    data = np.asarray(data)
    N = data.shape[0]
    if N == 0:
        return np.nan
    logdata = np.log(data)

    if lmb == 0:
        logvar = np.log(np.var(logdata, axis=0))
    else:
        logx = lmb * logdata - np.log(abs(lmb))
        logvar = _log_var(logx)

    return (lmb - 1) * np.sum(logdata, axis=0) - N/2 * logvar


def boxcox_llf_log_space_new(lmb, data):
    data = np.asarray(data)
    N = data.shape[0]
    if N == 0:
        return np.nan
    logdata = np.log(data)

    if lmb == 0:
        logvar = np.log(np.var(logdata, axis=0))
    else:
        logx = lmb * logdata
        logvar = _log_var(logx) - 2 * np.log(abs(lmb))

    return (lmb - 1) * np.sum(logdata, axis=0) - N/2 * logvar

x = [2003, 1950, 1997, 2000, 2009]

lmbs = np.linspace(-1e-5, 1e-5, 100)

llf_linear = [boxcox_llf_linear_space(l, x) for l in lmbs]
llf_log_old = [boxcox_llf_log_space_old(l, x) for l in lmbs]
llf_log_new = [boxcox_llf_log_space_new(l, x) for l in lmbs]

fig_linear, ax_linear = plt.subplots()
fig_log_old, ax_log_old = plt.subplots()
fig_log_new, ax_log_new = plt.subplots()

ax_linear.set_title('linear space')
ax_log_old.set_title('log space (old)')
ax_log_new.set_title('log space (new)')

ax_linear.plot(lmbs, llf_linear)
ax_log_old.plot(lmbs, llf_log_old)
ax_log_new.plot(lmbs, llf_log_new)

ylim = ax_log_old.get_ylim()
ax_linear.set_ylim(ylim)
ax_log_new.set_ylim(ylim)

fig_linear.savefig('linear.png', bbox_inches='tight', dpi=600)
fig_log_old.savefig('log-old.png', bbox_inches='tight', dpi=600)
fig_log_new.savefig('log-new.png', bbox_inches='tight', dpi=600)
```

</details>

### CODE REVIEW

**Comment by mdhaber** [comment]
Thanks. Please add a test (e.g. `test_instability_gh20021`) that fails with `main` and passes in this PR.

**Comment by xuefeng-xu** [comment]
The expected value was computed with `boxcox_llf` in `mpsci`.

<div align=center>
    
</div>

File: scipy/stats/tests/test_morestats.py [hunk 1]
```diff
@@ -1813,6 +1813,12 @@ def test_gh_6873(self):
         # The expected value was computed with mpmath.
         assert_allclose(llf, -17.93934208579061)
 
+    def test_instability_gh20021(self):
+        data = [2003, 1950, 1997, 2000, 2009]
+        llf = stats.boxcox_llf(1e-8, data)
+        # The expected value was computed with mpmath.
```
Comment (mdhaber) [inline scipy/stats/tests/test_morestats.py:1819]: I see that this was following the example above, but we now try to include enough information to reproduce reference values. Without it, it's harder for me to check the value so I can sign off on it, and it's even tougher for someone else who's interested in the future.

It's OK with me to use `mpsci` here, but can you include the essential informaiton, like:
```
# The expected value was computed with mpsci
# mpsci.stats.boxcox_llf(1e-8, data)
```
and maybe the `mpmath.mp.dps` value, if you set it? When I use 100 digits of precision, I got a different value with:

<details>

```python3
from mpmath import mp
mp.dps = 100

def var(x, ddof=0):
    """
    Variance of the values in the sequence x.
    """
    n = len(x)
    with mp.extraprec(16):
        sumx = mp.fsum(x)
        meanx = sumx / n
        varx = mp.fsum((mp.mpf(t) - meanx)**2 for t in x)/(n - ddof)
    return varx

def boxcox_llf(lam, x):
    """
    Log-likelihood for maximum likelihood estimation of the Box-Cox parameters.
    """
    with mp.extradps(5):
        lam = mp.mpf(lam)
        x = [mp.mpf(t) for t in x]
        n = len(x)

        logdata = [mp.log(t) for t in x]
        sumlogdata = mp.fsum(logdata)

        # Compute the variance of the transformed data.
        if lam == 0:
            variance = var(logdata)
        else:
            # Transform without the constant offset 1/lam.  The offset does
            # not effect the variance, and the subtraction of the offset can
            # lead to loss of precision.
            variance = var([t**lam / lam for t in x])

        return (lam - 1) * sumlogdata - n/2 * mp.log(variance)

data = [2003, 1950, 1997, 2000, 2009]
boxcox_llf(mp.mpf(1e-8), data)  # -15.3240127286901659
```

</details>

Comment (xuefeng-xu) [reply scipy/stats/tests/test_morestats.py]: Thanks for your suggestion, I add some comment and update the math above.

Comment (mdhaber) [reply scipy/stats/tests/test_morestats.py]: Thanks!

**Pull request merged**